ledger-api/grpc-definitions/com/daml/ledger/api/v1/transaction_filter.proto

```diff
@@ -12,30 +12,62 @@ option java_outer_classname = "TransactionFilterOuterClass";
 option java_package = "com.daml.ledger.api.v1";
 option csharp_namespace = "Com.Daml.Ledger.Api.V1";
 
-// Used for filtering Transaction and Active Contract Set streams.
-// Determines which on-ledger events will be served to the client.
+// A filter both for filtering create and archive events as well as for
+// filtering transaction trees.
 message TransactionFilter {
 
-  // Keys of the map determine which parties' on-ledger transactions are being queried.
-  // Values of the map determine which events are disclosed in the stream per party.
-  // At the minimum, a party needs to set an empty Filters message to receive any events.
   // Each key must be a valid PartyIdString (as described in ``value.proto``).
+  // The interpretation of the filter depends on the stream being filtered:
+  // (1) For **transaction tree streams** only party filters with wildcards are allowed, and all subtrees
+  //     whose root has one of the listed parties as an informee are returned.
+  // (2) For **transaction and active-contract-set streams** create and archive events are returned for all contracts whose
+  //    stakeholders include at least one of the listed parties and match the
+  //    per-party filter.
   // Required
   map<string, Filters> filters_by_party = 1;
 }
 
+// The union of a set of contract filters, or a wildcard.
 message Filters {
 
-  // If not set, no filters will be applied.
+  // If set, then contracts matching any of the ``InclusiveFilters`` match
+  // this filter.
+  // If not set, any contract matches this filter.
   // Optional
   InclusiveFilters inclusive = 1;
 }
 
-// If no internal fields are set, no filters will be applied.
+// A filter that matches all contracts that are either an instance of one of
+// the ``template_ids`` or that match one of the ``interface_filters``.
 message InclusiveFilters {
 
-  // A collection of templates.
+  // A collection of templates for which the payload will be included in the
+  // ``create_arguments`` of a matching ``CreatedEvent``.
   // SHOULD NOT contain duplicates.
-  // Required
+  // Optional
   repeated Identifier template_ids = 1;
+
+  // Include an ``InterfaceView`` for every ``InterfaceFilter`` matching a contract.
+  // Optional
```

Comment:
What happens if there are >1 `InterfaceFilter`s with the same `interface_id`?

Reply:
Good catch. @cocreature @rautenrieth-da : I'd suggest we require the `interface_id` of different interface filters to be unique; and throw a `INVALID_ARGUMENT` if that's not the case. What do you think?

Reply:
sounds sensible, I slightly dislike that this is inconsistent with template ids where afaik we dedup server side but it still seems better than trying to be clever here and do some merging of the fields set in each view.

Reply:
fixed

Reply:
The [PoC implementation](https://github.com/digital-asset/daml/pull/14141/files#diff-b13ee5abea9d4bec6d0df336956420ee16e57b1e8fb31c8efefbd0db5fe05b3fR83) would (I think) include the interface view if at least one of the `InterfaceFilters` had `include_interface_view` set, which sounds like sensible behavior to me.

Throwing an exception is also good as it likely catches a user error, but I agree with Moritz that we then maybe should throw the same exception for duplicate template ids?